### PR TITLE
Simpler ctxt table malloc scheme.

### PR DESCRIPTION
**Pull request by linas**
This simplifies the ideas from #1023 -- @ampli any comments?

### CODE REVIEW

**Comment by ampli** [comment]
In my fix, I only kept the biggest allocation (128MB) because only it had a bad performance affect.
But on Linux at least, `mmap()` is used by default from 32MB, and maybe other CPU types would be affected by 32MB and 64MB blocks too. So it is a good idea to keep every size as you did.

Also, your fix adds small efficiency in not needed to allocate/free this table at all over sentences, even 
if all of them are small (and hence the table allocation doesn't use `mmap()`). In general, a significant speedup could be achieved if we do it for other big data structures. One easy way is to free the memory pools only at the end of the program.

My original implementation of memory pools supported that (by creating a list of pools, to be freed on program exit), but this has been removed because it seemed to me too radical at the time. I can restore this implementation. The problem is that it may hold hundreds of MB. This can be mitigated by using the `memory` parse_option as a soft memory limit, and after each sentence memory pools from the said list of pools will get freed until this limit is satisfied.

**Comment by linas** [comment]
Should I merge this?

**Comment by ampli** [comment]
I think that not, since it never freeing the table, while our convention is to free everything upon exit.

**Comment by ampli** [comment]
I incorporated this idea into the dynamic sizing connector table patch (to be sent after the pending PRs are applied).
But a problem remains: **It causes a thread memory leak.**
Possible solutions:
1. Provide a function to be called any time after a thread starts, that will free this memory on thread end.
2. Revive the `max_memory` parse-options setup, with a new meaning of trimming the unused allocated memory to a requested value. The idea is that it can be called with zero before exiting a thread. Its usage may have other benefits, like being able to keep much of the memory allocated between sentences, yet provide API to limit such kept memory (which can be multiple gigabytes due to a single "problematic" sentence). It can also be used to prevent a too-big connector table on memory-limited systems.

**Comment by ampli** [comment]
The idea of this patch is incorporated in a yet-to-be-sent PR so I'm closing this PR.

**Comment by ampli** [comment]
Note my comment above about a remaining **thread memory leak** and my proposed solutions.
So I'm leaving it open after all for discussing that.

**Review by ampli — COMMENTED** [review]
Not to be applied, see below.

File: link-grammar/parse/count.c [hunk 1]
```diff
@@ -57,40 +57,43 @@ struct count_context_s
...
+	static TLS unsigned int log2_table_size = 0;
 
 	if (ctxt == NULL)
 	{
-		free(kept_table);
+		if (kept_table) free(kept_table);
+		kept_table = NULL;
+		log2_table_size = 0;
 		return;
 	}
```
[the remark this answers is not in the document]
Comment (linas) [reply link-grammar/parse/count.c:96]: > It seems this is unreachable, as nothing calls init_table(NULL, ...).

See above, line 81. you have to `#define DO_THE_ATEXIT_FREE 1` to call it.

**Comment by linas** [comment]
Using `max_memory` seems like a reasonable solution ...

**Comment by ampli** [comment]
Implemented in PR #1125.

**Pull request closed without merging**